src/query/service/src/pipelines/processors/transforms/transform_resort_addon_without_source_schema.rs

```diff
@@ -0,0 +1,129 @@
+
+use std::sync::Arc;
+
+use common_catalog::table_context::TableContext;
+use common_exception::Result;
+use common_expression::BlockMetaInfoDowncast;
+use common_expression::DataBlock;
+use common_expression::DataSchemaRef;
+use common_expression::Expr;
+use common_expression::Scalar;
+use common_sql::evaluator::BlockOperator;
+use common_sql::evaluator::CompoundBlockOperator;
+use common_sql::parse_exprs;
+use common_storages_factory::Table;
+
+use crate::pipelines::processors::port::InputPort;
+use crate::pipelines::processors::port::OutputPort;
+use crate::pipelines::processors::processor::ProcessorPtr;
+use crate::pipelines::processors::transforms::transform::Transform;
+use crate::pipelines::processors::transforms::transform::Transformer;
+use crate::sessions::QueryContext;
+
+pub struct TransformResortAddOnWithoutSourceSchema {
+    output_schema: DataSchemaRef,
+    ctx: Arc<QueryContext>,
+    table: Arc<dyn Table>,
+}
+
+impl TransformResortAddOnWithoutSourceSchema {
+    fn build_expression_transform(
+        &self,
+        input_schema: DataSchemaRef,
+    ) -> Result<CompoundBlockOperator> {
+        let mut exprs = Vec::with_capacity(self.output_schema.fields().len());
+        for f in self.output_schema.fields().iter() {
+            let expr = if !input_schema.has_field(f.name()) {
+                if let Some(default_expr) = f.default_expr() {
+                    let mut expr = parse_exprs(self.ctx.clone(), self.table.clone(), default_expr)?;
+                    let mut expr = expr.remove(0);
+                    if expr.data_type() != f.data_type() {
+                        expr = Expr::Cast {
+                            span: None,
+                            is_try: f.data_type().is_nullable(),
+                            expr: Box::new(expr),
+                            dest_type: f.data_type().clone(),
+                        };
+                    }
+                    expr
+                } else {
+                    let default_value = Scalar::default_value(f.data_type());
+                    Expr::Constant {
+                        span: None,
+                        scalar: default_value,
+                        data_type: f.data_type().clone(),
+                    }
+                }
+            } else {
+                let field = input_schema.field_with_name(f.name()).unwrap();
+                let id = input_schema.index_of(f.name()).unwrap();
+                Expr::ColumnRef {
+                    span: None,
+                    id,
+                    data_type: field.data_type().clone(),
+                    display_name: field.name().clone(),
+                }
+            };
+            exprs.push(expr);
+        }
+
+        let func_ctx = self.ctx.get_function_context()?;
+        Ok(CompoundBlockOperator {
+            ctx: func_ctx,
+            operators: vec![BlockOperator::Map {
+                exprs,
+                projections: None,
+            }],
+        })
```

Comment:
seems there are duplications with `TransformResortAddOn`?  can we do a refactoring by the way?

Reply:
No, transformAddOn need specify the source schema, but for merge into, the insert can be various, not consistent